### ISSUE TITLE
Slow inference when processing less than X amount of images per second

### ISSUE THREAD

**Issue by jordyvessen** [ultralytics/ultralytics#18494]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I was testing my YoloV8 model with a camera and noticed something strange when adjusting the camera's FPS. The higher the FPS of the camera, the faster the inference speed became. First I thought it was related to the GPU going to idle in between frames but when I used the CPU for inference the issue persisted.

The code below results in a fast inference time with an average of about ~8ms. The time is based of the standard Yolo output. I also tried using custom duration measurement and the torch.profiler but both resulted in similar durations.
```python
# Python 3.10.16
import cv2 as cv # 4.10.0
import torch # 2.4.1+cu121
import torchvision # 0.19.1+cu121
import time as tm

from ultralytics import YOLO # 8.3.56

if __name__ == '__main__':
    image_path = 'image.jpg'
    model = YOLO("yolov8m.pt")
    device = torch.device('cuda:0')
    model.to(device)
    
    src = cv.imread(image_path)
    for i in range(100):
        # tm.sleep(0.5)
        with torch.no_grad():
            results = model.predict(src, device=device)
```

When enabling the delay of half a second the duration increases to about ~16ms.
The same happens when using the 'cpu' to run inference, but then the time difference is 120vs140 ms.

Does anyone have an idea what is going on here? Or any suggestions on how to better diagnose the issue?

**Hardware**
Configuration 1:
Windows 10
CPU: I7 12700KF
GPU: RTX 4090
RAM: 64GB 4800MHz

NVIDIA-SMI 560.94                 
Driver Version: 560.94         
CUDA Version: 12.6

Configuration 2: (Timings are different but the issue is the same)
Windows 10
CPU: I7 13700T
GPU: RTX 4070
RAM: 32GB 4800Mhz

Configuration 3: (Laptop, the issue is not reproducible) 
Windows 11
CPU: Ultra 7 165H
GPU: RTX 2000 Ada laptop gpu
RAM: 64GB 5600 MT/s

NVIDIA-SMI 556.35                 
Driver Version: 556.35         
CUDA Version: 12.5

**Image**
![image](https://github.com/user-attachments/assets/8ec41d68-4184-4b90-8888-8cd3adf96d2b)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jordyvessen, thank you for your detailed report and for using Ultralytics 🚀! Your observations regarding inference speed are interesting, and we’d love to investigate this further.

For now, here are some steps and resources that might help us better understand and resolve the issue:

### 🐛 Reproducibility
It appears you've provided a solid code example—thank you! If possible, could you simplify this into a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum_reproducible_example/)? For instance:
1. Share how the delay at `tm.sleep(0.5)` impacts inference speed.
2. Share any console output or logs that might provide further insights into performance differences.

### ✅ System and Environments
To ensure the latest improvements and bug fixes are included, please verify that you're using the most up-to-date version of Ultralytics and related dependencies. You can upgrade your library using the following command:
```bash
pip install -U ultralytics
```
Also, please ensure your environment matches [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) and is running Python>=3.8 with PyTorch>=1.8.

If the issue persists, try testing inference in any of our verified environments to compare performance:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/).
- **Amazon AWS**: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).
- **Docker Image**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.

### 💡 General Recommendations
- Verify if the increased inference time with delays is caused by GPU or CPU runtime idling.
- Try running the model in a clean environment and compare results (e.g., Google Colab).
- Experiment with `torch.set_grad_enabled(False)` to rule out autograd overheads explicitly.

### 🌍 Community and Support
If you have additional findings, feel free to share them here! You can also check out and discuss with our active community across different platforms:
- [Discussions](https://github.com/ultralytics/ultralytics/discussions): In-depth Q&A.
- [Discord](https://discord.com/invite/ultralytics): Real-time chat 🎧.
- [Subreddit](https://reddit.com/r/Ultralytics): Discuss your findings with other users.

### 🔄 Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>  
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing, meaning the issue may be environment-specific.

---

This is an automated response, but an Ultralytics engineer will review and assist you further shortly! 😊

**Comment by Y-T-G** [comment]
None of these should be used with Ultralytics:
```
device = torch.device('cuda:0')
model.to(device)   
with torch.no_grad():
```

You can try exporting it to a different format like TensorRT and see if the issue exists:
https://docs.ultralytics.com/modes/export/

**Comment by jordyvessen** [comment]
Thanks for your input, but without those lines the issue persists.
I made a colab which reprocduces the issue.

https://colab.research.google.com/drive/1iblXOEsGj_6wSqZpdzDoaHtUTV-d132H?usp=sharing

Below a screenshot of the results without the delay
![image](https://github.com/user-attachments/assets/8a23624b-e630-48c3-96cc-55c58970e864)

And with the delay:
![image](https://github.com/user-attachments/assets/91784b08-4e57-4205-8758-ae4057b49f62)

**Comment by Y-T-G** [comment]
Can you try exporting to TensorRT?

**Comment by Y-T-G** [comment]
I Know TensorRT has this issue

https://forums.developer.nvidia.com/t/tensorrt-engine-inference-faster-when-execute-is-called-more-often/202570

https://forums.developer.nvidia.com/t/inference-time-becomes-longer-when-doing-non-continuous-fp16-or-int8-inference/184127

But I'm not sure if it's an Ultralytics issue or something low-level as to how scheduling is done for processes.

**Comment by Y-T-G** [comment]
https://discuss.pytorch.org/t/inconsistent-model-inference-time/106933

https://github.com/microsoft/onnxruntime/issues/7449
